### PR TITLE
New kernel lr

### PR DESCRIPTION
**Pull request by meyerscetbon**

### CODE REVIEW

**Comment by michalk8** [comment]
> https://github.com/ott-jax/ott/pull/299#issuecomment-1449528547

Thanks Meyer! Is there a reason why kernel mode cannot handle 0 weights (e.g., division by 0)?

**Comment by meyerscetbon** [comment]
Because you need to specify the support of the measures before performing the updates of u1 and u2 in the Dykstra algorithm as you might have 0 / 0 type operations.

**Comment by marcocuturi** [comment]
> Because you need to specify the support of the measures before performing the updates of u1 and u2 in the Dykstra algorithm as you might have 0 / 0 type operations.

I think this should be handled then, as done in other `sinkhorn` operations, for instance:
https://github.com/ott-jax/ott/blob/2abdd72de5545d169a00d8aba6966393a86d3480/src/ott/solvers/linear/sinkhorn.py#L252

Handling 0 weights is important for padding measures with 0's to get faster execution for measures with variable sizes.

File: src/ott/solvers/linear/sinkhorn_lr.py [hunk 1]
```diff
@@ -292,7 +281,7 @@ def __init__(
       kwargs_init: Optional[Mapping[str, Any]] = None,
       **kwargs: Any,
   ):
-    assert lse_mode, "Kernel mode not yet implemented."
+    # assert lse_mode, "Kernel mode not yet implemented."
```
Comment (marcocuturi) [inline src/ott/solvers/linear/sinkhorn_lr.py:284]: please remove

File: src/ott/solvers/linear/sinkhorn_lr.py [hunk 2]
```diff
@@ -492,24 +513,123 @@ def recompute_couplings(
...
+    def body_fn(
+        iteration: int, constants: Tuple[jnp.ndarray, ...],
+        state_inner: Tuple[jnp.ndarray, ...], compute_error: bool
+    ) -> Tuple[jnp.ndarray, ...]:
+      # TODO(michalk8): in the future, use `NamedTuple`
+      u1, u2, v1_old, v2_old, g_old, q_gi, q_gp, q_q, q_r, err = state_inner
+      k_q, k_r, k_g, a, b = constants
+
+      # First Projection
+      u1 = a / jnp.dot(k_q, v1_old)
```
Comment (marcocuturi) [inline src/ott/solvers/linear/sinkhorn_lr.py:562]: would 
```
supp_a = a > 0
```
stored before the loop, and then
```
u1 = jnp.where(a >0, a / jnp.dot(k_q, v1_old), 0.0)
```

(and same for `u2`) help you get rid of the issues you had with tests when some elements in `a` or `b` are 0?

**Review by marcocuturi — CHANGES_REQUESTED** [review]

File: src/ott/solvers/linear/sinkhorn_lr.py (hunk 1 above)
Comment (michalk8) [reply src/ott/solvers/linear/sinkhorn_lr.py]: Done.

File: src/ott/solvers/linear/sinkhorn_lr.py (hunk 2 above)
Comment (michalk8) [reply src/ott/solvers/linear/sinkhorn_lr.py:581]: This wouldn't have helped (I think), since we're not dividing by `0`. The issue was just convergence, `epsilon=0.5, gamma=10`, which was high `gamma`, lowered it to `1.0` and locally it passes.

Comment (michalk8) [reply src/ott/solvers/linear/sinkhorn_lr.py]: This solved the issue!

**Review by michalk8 — APPROVED** [review]
